### PR TITLE
update error message when output can't be pickled by fs_io_manager

### PR DESCRIPTION
**Pull request by jamiedemaria**
Catches when an output can't be pickled by the fs_io_manager and raises a more helpful error message (#5551)

Pickle is a bit ambiguous about what Exceptions could be thrown by pickle.dumps()
From the docs `pickle.PicklingError` will be thrown when an object is unpicklable, `RecursionError` will be thrown when trying to pickle recursive structures that exceed the python recursion limit. For `AttributeError` and `ImportError` the docs state:
"""
Note that functions (built-in and user-defined) are pickled by "fully qualified" name reference, not by value. This means that only the function name is pickled, along with the name of the module the function is defined in. Neither the function's code, nor any of its function attributes are pickled. Thus the defining module must be importable in the unpickling environment, and the module must contain the named object, otherwise an exception will be raised. [3]

[3] The exception raised will likely be an ImportError or an AttributeError but it could be something else.
"""

This leaves it open that another error could be thrown if an object isn't pickleable 

This PR manages `AttributeError` `ImportError` `RecursionError` and `pickle.PicklingError` but if one of the "something else" exceptions is thrown by pickle, the original error will be raised to the user. 

Unfortunately, `pickle` doesn't provide a function to check if an object is picklable. the library `dill` does, but `dill` is capable of pickling more types of objects than `pickle` and can only tell us if `dill` can pickle the object. So first checking if `dill` says an object is picklable wouldn't help us if we then use `pickle.dumps()`. 

Definitely open to suggestions on other ways to  handle this error

### CODE REVIEW

**Comment by dpeng817** [comment]
cc @alangenfeld I'm wondering if you have thoughts on the module-scoped function pickling dilemma. It seems like it would be an error raised when attempting to re-load the fxn into memory, ie in load_input. The only scenario I can imagine this happening is re-execution when you try to re-load the pickled fxn, and it no longer exists in module scope.

File: python_modules/dagster/dagster/core/storage/fs_io_manager.py [hunk 1]
```diff
@@ -119,7 +120,20 @@ def handle_output(self, context, obj):
...
+                executor = context.step_context.pipeline_def.mode_definitions[0].executor_defs[0]
+
+                raise DagsterInvariantViolationError(
+                    f"Object {obj} is not picklable. Use the mem_io_manager with an in process "
+                    "executor to avoid pickling outputs. You are currently using the "
+                    f"fs_io_manager and the {executor.name}. \n"
+                    "For more information on io managers, visit "
+                    "https://docs.dagster.io/concepts/io-management/io-managers \n"
+                    "For more information on executors, vist "
+                    "https://docs.dagster.io/deployment/executors#overview"
```
Comment (dpeng817) [inline python_modules/dagster/dagster/core/storage/fs_io_manager.py:135]: This is an awesome error message

File: python_modules/dagster/dagster_tests/core_tests/storage_tests/test_fs_io_manager.py [hunk 2]
```diff
@@ -100,3 +102,35 @@ def get_solid_version(self, _):
             result = my_job.execute_in_process(instance=instance)
             assert result.success
             assert len(recorder) == 1
+
+
+def test_fs_io_manager_unpicklable():
+    @op
+    def unpicklable_output():
+        # lambda functions can't be pickled
+        return lambda x: x*x
```
Comment (dpeng817) [inline python_modules/dagster/dagster_tests/core_tests/storage_tests/test_fs_io_manager.py:111]: probably worth testing the full gamut of pickle errors that we can get, ie recursion error, direct unpickleable error, etc

File: python_modules/dagster/dagster/core/storage/fs_io_manager.py (hunk 1 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/core/storage/fs_io_manager.py]: I shifted the language a bit to suggest the in memory and in process as a potential option. lmk how it reads

File: python_modules/dagster/dagster_tests/core_tests/storage_tests/test_fs_io_manager.py (hunk 2 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster_tests/core_tests/storage_tests/test_fs_io_manager.py]: yup - added them in

File: python_modules/dagster/dagster/core/storage/fs_io_manager.py [hunk 3]
```diff
@@ -119,7 +120,28 @@ def handle_output(self, context, obj):
...
+                    # throw a RecursionError
+                    obj_repr = f"{obj.__class__} exceeds recursion limit and"
+                else:
+                    obj_repr = obj.__str__()
+
+                raise DagsterInvariantViolationError(
+                    f"Object {obj_repr} is not picklable. You are currently using the "
+                    f"fs_io_manager and the {executor.name}. You will need to use a different "
+                    "io manager to continue using this output. For example, you can use the "
+                    "mem_io_manager with an in process executor.\n"
```
Comment (dpeng817) [inline python_modules/dagster/dagster/core/storage/fs_io_manager.py:139]: nit: `an in process executor` -> `the inprocess executor`

**Review by alangenfeld — APPROVED** [review]
looks good to me!

**Pull request merged**